File: build_hamiltonian.py
from modular_arithmetic import circ_shift, to_base
from basis_construction import get_momentum_projection, get_period, check_momentum, get_sector_basis, construct_period_table, construct_momentum_table, get_Zn_charge
import numpy as np
from numpy import kron
from scipy.linalg import expm, eigh
from spin_ops import Sx,Sy,Sz, Id
import logging 

def find_index(num, num_list):
    # Binary search for num in sorted num_list, return index or -1 if not found
    left, right = 0, len(num_list) - 1
    while left <= right:
        mid = (left + right) // 2
        if num_list[mid] == num:
            return mid
        elif num_list[mid] < num:
            left = mid + 1
        else:
            right = mid - 1
    return None

def build_nn_hamiltonian_nosym(H,nsites):
    # Build periodic hamiltonian from nearest neighbor `H` on `nsites`
    hib_onsite = round(np.sqrt(H.shape[0]))
    assert H.shape == (hib_onsite*hib_onsite, hib_onsite*hib_onsite), f"hib_onsite ({hib_onsite}) does not match that of the shape of H ({H.shape})"
    assert nsites >= 2, "Nearest Neighbour Hamiltonian needs at least 2 sites"

    # Build hamiltonian by direct product 
    H0 = H 
    for i in range(nsites-2):
        H0 = kron(H0,np.eye(hib_onsite))
    H0 = H0.reshape([hib_onsite]*(2*nsites)) # relabel the legs 
    print('H0 has shape', H0.shape)

    # Cycle through all sites and sum the hamiltonian 
    idcs = np.arange(nsites)
    #print(idcs)
    permute_axes = [np.concatenate(((idcs+i) % nsites,(idcs+i) % nsites+nsites)) for i in range(nsites)]
    H_periodic = sum([np.permute_dims(H0,ax) for ax in permute_axes])
    return H_periodic.reshape(hib_onsite**nsites,hib_onsite**nsites)


def apply_nn_hamiltonian(H,m,a,nsites,base=3):
    # Apply hamiltonian `H(m+1,m)`` to the state `a` periodically, with a number of sites `nsites`
    # The site indicies goes from `0` to `nsites-1`, which has a physical layout from right to left
    hib_size = base**2 
    assert H.shape == (hib_size,hib_size), f"Hamiltonian shape {H.shape} does not match that of the hilbert space {(hib_size, hib_size)}"

    # Get the shifted_states, it's head, and it's tail 
    state_shifted = circ_shift(a,-m,base,nsites) # When m = 0, we don't need to shift the states
    state_tail = state_shifted % base**2 # Get the two qudit tail of the state
    state_head = state_shifted - state_tail

    # Get the Hamiltonian coefficients 
    coeff = H[:,state_tail]

    # Get the shifted states 
    b = [circ_shift(state_head + i,m,base,nsites) for i in range(hib_size)]

    return coeff, b # the coefficients, the states

def build_nn_hamiltonian_charged_sector(H,nsites,charged_states,momentum,charge,charge_checker,base=3):
    # Float type 
    float_type = np.complex128
    # Compute crystal momentum 
    k = 2*np.pi *momentum/nsites

    # Build the matrix 
    sector_size = len(charged_states)
    logging.debug(f"Building k={momentum} sector with size {sector_size}x{sector_size}")
    H_sector = np.zeros(shape=(sector_size,sector_size),dtype=float_type)
    
    for a in charged_states:
        Pa = get_period(a,nsites,base)
        for m in range(nsites):
            coeffs, bs = apply_nn_hamiltonian(H,m,a,nsites,base)
            for coeff, b in zip(coeffs,bs):
                b_rep, Pb, b_r = get_momentum_projection(b,nsites,momentum,base)

                if b_rep >= 0 and charge_checker(b,charge): # Only if compatible 
                    assert b_rep in charged_states, f"The state {b_rep} is lot in the list of states {charged_states}"
                    phase = np.exp(-1.j*k*b_r)
                    norm = np.sqrt(Pa/Pb)
                    logging.debug(f"<{to_base(b_rep,base,nsites)}|H_{m}|{to_base(a,base,nsites)}> += {norm*phase*coeff}")
                    idc_out, idc_in = find_index(b_rep,charged_states), find_index(a,charged_states)
                    H_sector[idc_out,idc_in] += norm*phase*coeff
    
    atol = 100*np.finfo(float_type).eps
    if not np.allclose(H_sector, H_sector.conj().T, atol=atol):
        max_diff = np.max(np.abs(H_sector - H_sector.conj().T))
        raise AssertionError(f"H_sector is not Hermitian up to 100 times machine precision {atol}. Max difference: {max_diff}")
    return H_sector

# def build_nn_hamiltonian_sector(H,momentum,momentum_table)
# logging.basicConfig(level=logging.DEBUG)
if __name__ == "__main__":
    nsites = 4
    
    # Period table 
    pt = construct_period_table(nsites)
    mt = construct_momentum_table(pt)
    print(pt)
    print(mt)

    # Build random matrices to test the code
    O1 = np.random.rand(3,3) + 1j * np.random.rand(3,3)
    O1 = 0.5 * (O1 + O1.conj().T)
    O2 = np.random.rand(3,3) + 1j * np.random.rand(3,3)
    O2 = 0.5 * (O2 + O2.conj().T)
    Hnn = kron(O1,O2)#kron(Sx,Sx)+kron(Sy,Sy)

    for momentum in range(0,nsites):
        k_checker = lambda state, momentum: check_momentum(state,momentum,nsites)
        sector_basis = get_sector_basis(nsites,(momentum,),(k_checker,))
        print(f'Sector basis for k={momentum}',sector_basis)
        H = build_nn_hamiltonian_charged_sector(Hnn,nsites,sector_basis,momentum,charge=-1,charge_checker=lambda s,c:True)
        #H = build_nn_hamiltonian_sector(Hnn, nsites, momentum, momentum_table, period_table)
        print("Sector size: ",H.shape)
        eigval = eigh(H, eigvals_only=True)
        eigval_str = ", ".join(f"{v:8.3f}" for v in eigval)
        print(f"k={momentum}:\n\t{eigval_str}")

    # Exact diagonalization without symmetry 
    H = build_nn_hamiltonian_nosym(Hnn,nsites)
    print("Sector size: ", H.shape)
    eigval = eigh(H,eigvals_only=True)
    eigval_str = ", ".join(f"{v:8.3f}" for v in eigval)
    print(f"No symmetry:\n{eigval_str}")


    # print("Checking that the state enumeration works")
    # state_str = "012"
    # nsites = len(state_str)
    # a = digits_str_to_num(state_str,base=3)
    # Rx = expm(1.j*np.pi*Sx)
    # H = kron(Id,Rx)
    # print('Rx matrix:')
    # for row in Rx:
    #     print('\t'.join(f'{elem.real:+.3f}{elem.imag:+.3f}j' for elem in row))
    # print("Input states: ",state_str)
    # coeff, out_states = apply_nn_hamiltonian(H=H,m=-1,a=a,nsites=nsites)
    # print("Possible out states are when applying to m=-1:")
    # for c,b in zip(coeff,out_states):
    #     print(to_base(b,base=3,ndigit=nsites), 'coeff:', c)        

    # print()
    # coeff, out_states = apply_nn_hamiltonian(H=H,m=0,a=a,nsites=nsites)
    # print("Possible out states are when applying to m=0:")
    # for c,b in zip(coeff,out_states):
    #     print(to_base(b,base=3,ndigit=nsites), 'coeff:', c) 

    # print()
    # coeff, out_states = apply_nn_hamiltonian(H=H,m=1,a=a,nsites=nsites)
    # print("Possible out states are when applying to m=1:")
    # for c,b in zip(coeff,out_states):
    #     print(to_base(b,base=3,ndigit=nsites), 'coeff:', c) 

File: julia/build_hamiltonian.jl
include("./modular_arithmetic.jl")
include("./basis_construction.jl")

function get_momentum_projection(s::Int,nsites::Int,momentum::Int,base::Int)
    sᵢ = s
    smin = s
    R_to_min = 0

    R = 1 
    while R <= nsites
        sᵢ = circ_shift(sᵢ,1,;base=base,ndit=nsites)
        if sᵢ  < smin
            smin = sᵢ
            R_to_min = R 
        end 
        if sᵢ == s
            break 
        end
        R += 1
    end 
    # Check momentum compatibility
    (momentum*R) %nsites != 0 ? (-1,-1,-1) : (smin,R,R_to_min)
end

function apply_nn_hamiltonian(;H::Matrix{T}, site::Int,state_in::Int,nsites::Int, base::Int) where T <: Number
    hib_size = base^2 
    size(H) == (hib_size,hib_size) || throw("Hamiltonian shape $(size(H)) does not match that of the Hilbert space.")

    # Get the shifted_states
    state_shifted = circ_shift(state_in,-site;base=base,ndit=nsites)
    state_tail = mod(state_shifted, hib_size)
    state_head = state_shifted - state_tail 

    # Get the Hamiltonian coefficients 
    coeff = H[1:end,state_tail + 1] # Julia index start from 1
    
    # Get the shifted states 
    state_out = [circ_shift(state_head+i,site;base=base,ndit=nsites) for i in 0:(hib_size-1)]

    return coeff, state_out
end

# Hamiltonian without momentum 
struct NosymHamiltonianNN{T<: Number}
    Hnn::Matrix{T}
    nsites::Int
    base::Int
end

function (H::NosymHamiltonianNN{T})(v0::AbstractVector{T}) where {T <: Number}
    
    nsites = H.nsites
    base = H.base
    Hnn = H.Hnn
    Hv = zeros(T,base^nsites)

    for i in 1:length(v0)
        for site in 0:(nsites-1)
            state_in = i-1
            coeffs, bs = apply_nn_hamiltonian(;H=Hnn,site=site,state_in=state_in,nsites=nsites,base=base)
            for (c, b) in zip(coeffs, bs)
                Hv[b+1] += c*v0[i]
            end
        end
    end
    return Hv
end

block_size(H::NosymHamiltonianNN) = H.base^H.nsites
function dtype(H::NosymHamiltonianNN{T}) where {T}
    return T
end
function get_matrix(H::NosymHamiltonianNN{T}) where {T <: Number}
    dim = block_size(H)
    M = zeros(T,dim,dim)
    for i in 1:dim 
        e = zeros(T,dim)
        e[i] = one(T)
        M[1:end,i] = H(e)
    end
    return M
end

function get_nn_hamiltonian_nosym(hnn::Matrix{T},nsites::Int,base::Int) where {T <: Number}
    Id = Matrix{T}(I,base,base)
    H0 = hnn
    for _ in 1:(nsites-2)
        H0 = kron(H0,Id)
    end
    # Reshape into 2*nsites tensor 
    dims = ntuple(_-> base,2*nsites)
    H0 = reshape(H0,dims)

    # Build periodic sum 
    H_periodic = zero(H0)
    idcs = collect(1:nsites)

    for shift in 0:(nsites-1)

        left  = ((idcs .+ shift .- 1) .% nsites) .+ 1
        right = left .+ nsites

        perm = vcat(left, right)

        H_periodic .+= permutedims(H0, perm)
    end

    # ---- Reshape back to matrix ----
    dim = base^nsites
    return reshape(H_periodic, dim, dim)
end
# Periodic Hamiltonian 
struct PeriodicHamiltonianNN{T<: Number}
    Hnn::Matrix{T}
    momentum_sector::MomentumSector
    _cache_srep::Dict{Int,Int} # Caching mechanism to speed up evaluation
    _cache_R::Dict{Int,Int}
    _cache_R_to_min::Dict{Int,Int}

    function PeriodicHamiltonianNN(Hnn::Matrix{T},ms::MomentumSector) where {T <: Number}
        new{T}(Hnn,ms,Dict{Int,Int}(),Dict{Int,Int}(),Dict{Int,Int}())
    end
end

function get_state_rep(H::PeriodicHamiltonianNN,state::Int)
    srep = get(H._cache_srep,state,-2)  # YM, 2/27/26. Tried false before. Apparently 0 == false is true so it kept recomputing the cache. So I use -2 instead.
    if srep == -2
        basis = H.momentum_sector
        nsites = basis.nsites
        momentum = basis.momentum.k
        base = basis.base
        srep, R, R_to_min = get_momentum_projection(state,nsites,momentum,base)
        H._cache_srep[state] = srep
        H._cache_R[state] = R 
        H._cache_R_to_min[state] = R_to_min
        # println("Computing cache $(length(H._cache_srep))")
        return (srep, R, R_to_min)
    end
    R = H._cache_R[state]
    R_to_min = H._cache_R_to_min[state]
    return (srep, R, R_to_min)
end

function block_size(H::PeriodicHamiltonianNN)
    return length(H.momentum_sector.state_list)
end

function dtype(H::PeriodicHamiltonianNN{T}) where {T}
    return T
end

function (H::PeriodicHamiltonianNN{T})(v0::AbstractVector{T}) where {T <: Number}
    sector_size = block_size(H)
    length(v0) == sector_size || throw("v0 has shape $(size(v0)), but there are $(sector_size) basis")

    Hv = zeros(T,sector_size)
    basis = H.momentum_sector
    nsites = basis.nsites
    base = basis.base
    momentum = basis.momentum.k # Momentum (integer)
    charges = basis.charges
    state_list = basis.state_list
    period_list = basis.period_list

    k = 2*π*momentum/nsites
    for i in 1:length(v0) # Loop over columns 
        ψin = state_list[i]
        Pin = period_list[i]
        for site in 0:(nsites-1) # H(site, site+1) for site = 0, ..., site-1; site+nsites ∼ site 
            coeffs, ψouts = apply_nn_hamiltonian(;H=H.Hnn,site=site,state_in=ψin,nsites=nsites,base=base)
            
            for (coeff,  ψout) in zip(coeffs, ψouts)
                # ψout_rep, Pout,  ψout_r = get_momentum_projection(ψout,nsites,momentum,base)
                ψout_rep, Pout,  ψout_r = get_state_rep(H,ψout) # YM, 2/27/2026. Optimization by caching. Much faster than before.
                # println("R_to_min: $(ψout_r); k=$k")
                if ψout_rep >= 0
                    idc_out = find_index(ψout_rep,basis)
                    if  idc_out > 0
                        phase = exp(-im*k*ψout_r)
                        norm = sqrt(Pin/Pout)
                        # println("H[$idc_out] += $(norm*phase*coeff)")
                        Hv[idc_out] += norm*phase*coeff*v0[i]
                    end
                end
            end
        end
    end
    return Hv
end

function get_matrix(H::PeriodicHamiltonianNN{T}) where {T <: Number}
    dim = block_size(H)
    M = zeros(T,dim,dim)
    for i in 1:dim 
        e = zeros(T,dim)
        e[i] = one(T)
        M[1:end,i] = H(e)
    end
    return M
end


